src/mon/AuthMonitor.cc

```diff
@@ -1791,6 +1734,172 @@ bool AuthMonitor::prepare_command(MonOpRequestRef op)
   return false;
 }
 
+/* Generate the caps that should now be present in the entity's auth keyring
+ * (that is wanted caps) by merging the caps already held in the auth keyring
+ * and the new caps provided by "fs authorize" subcommand.
+ */
+bool AuthMonitor::_gen_wanted_caps(EntityAuth& e_auth, vector<string>& newcaps,
+                                    string& fs)
+{
+  bool all_caps_same = true; // whether all caps passed are same as the
+			     // one present.
+
+  if (e_auth.caps.empty()) {
+    all_caps_same = false;
+    return all_caps_same;
+  }
+
+  for (auto it = newcaps.begin(); it != newcaps.end(); it += 2) {
+    string& cap_entity = *it;
+    string& ncap = *(it + 1);
+
+    if (e_auth.caps.count(cap_entity) == 0) {
+      all_caps_same = false;
+      continue;
+    }
+
+    string ecap; // ecap = entity cap
+    auto i = e_auth.caps[cap_entity].cbegin();
+    decode(ecap, i);
+    if (ecap == ncap) {
+      continue;
+    }
+
+    all_caps_same = false;
+
+    if (cap_entity == "mon") {
+      if (ecap.find("fsname=" + fs) == string::npos) {
```

Comment:
You're reimplementing cap parsing here. I'm not very familiar with this part of the Monitor but shouldn't it be possible to create an in-memory representation of the caps (like `MDSAuthCaps`), write a merge method for it (taking another `MDSAuthCaps`), and finally write its string representation out?

Reply:
instead of using raw string manipulation, would be better to expand `MDSCapParser` or define yet another parse using `boost::spirit`? so it's more readable and maintainable?